### PR TITLE
Initail Commit for NSCA plugin

### PR DESCRIPTION
**Pull request by IrappaPattar**
- [x] Create telegraf output plugin to send Linux system load information to Icinga1.
- [x] Initialise the NSCA telegraf output plugin.
- [x] Create a NSCA Client 
- [x] Connect to NSCA Server
- [x] Add Encryption Methods
- [x] Write data to NSCA Server

### CODE REVIEW

**Comment by drunlade** [comment]
@IrappaPattar Keep in mind you should almost never have a single person on a review.

**Comment by skldt** [comment]
@IrappaPattar 

1. Please add more information in the commit messages and PR description
2. Please also include a link to the issue you're working on
3. Have you used code form somewhere else as well? If yes, please link to that and mention in the commit message. (This is a huge amount of code for a couple of commits :) )

**Comment by IrappaPattar** [comment]
@skldt addressed to your comments. and not able to connect the PR with issue.

**Comment by skldt** [comment]
@IrappaPattar Could you specify what you need reviewed please? Like what code you've taken directly from another source and what are your modifications to it? (seems to be https://github.com/Syncbak-Git/nsca, but it'd be nice to have the exact revision as well)

I don't have the domain knowledge to review this, but I may be able to review your changes that are on top of existing code. Your changes should be a separate commit and the commit message for the existing code should link to where they've been taken from.

Do you think Martin would be in a better position to review this?

**Comment by IrappaPattar** [comment]
@skldt Martin don't know about golang. 

Please check whether this new NSCA plugin is as like other output plugins and this NSCA is newly created plugin.

Yes I have referred  the code from this https://github.com/Syncbak-Git/nsca.

**Comment by skldt** [comment]
@IrappaPattar I've reorganised the commits in this branch to make it easier to see your changes: https://github.com/thermeon/telegraf/commits/feature/413_initial_draft_organise_commits

In your changes, a couple of structs were moved (`NCSAServer` and `ServerInfo`) from their original positions in the code which made the diff harder to read. I've moved them back to their original positions for an easier to read diff.

Could you please check out this branch locally and add better commit messages for your code and commit it again? Also, could you please verify that my changes haven't changed the behaviour of your code?

Once you're done with that, you should be able to modify this branch to have those commits.

File: plugins/outputs/nsca/packet_test.go
Comment (skldt) [inline plugins/outputs/nsca/packet_test.go]: Why is this change required?

File: plugins/outputs/nsca/nsca.go
Comment (skldt) [inline plugins/outputs/nsca/nsca.go]: Why was this changed?

File: plugins/outputs/nsca/nsca.go
Comment (skldt) [inline plugins/outputs/nsca/nsca.go]: Why was this changed?

**Review by skldt — CHANGES_REQUESTED** [review]
Based on my limited knowledge, I'm not very keen on this design.

Here's what I think you've done - please correct me if I'm wrong:

1. Copied over the client code from https://github.com/Syncbak-Git/nsca manually . I'm assuming that the client code is independent of telgraf.
2. Updated the `NSCAServer` struct in the client code to implement methods that are required for a telegraf plugin
3. Changed signatures of existing methods of `NSCAServer` to match the interface for a telegraf plugin.

I don't think this is an ideal approach because:

1. We can't get updates from the upstream repository (https://github.com/Syncbak-Git/nsca) if we ever need it since the code is not in a `vendor` directory.
2. There's mixing of the NSCA client code and the telegraf plugin code (especially with the changing method signatures), but they can be written independently to be more maintainable.

I think this is what you should do:

1. Add the NSCA client package as a dependency of `telegraf`. See https://github.com/influxdata/telegraf/blob/master/CONTRIBUTING.md#adding-a-dependency 
2. Write a fresh plugin source file that includes an NSCA client in it and implements the interface needed for telegraf plugins.

That should hide the complexity of the NSCA client and also make your telegraf plugin implementation more clear.

**Comment by drunlade** [comment]
Well, the code looks sane (and I believe based on existing code anyway), I don't know enough about what it's up to to judge how right it is, so other than @skldt comments, looks fine to me.

**Comment by skldt** [comment]
@IrappaPattar 

Since the server to communicate with is [`NSCA-ng`](http://www.nsca-ng.org/) which uses a different communication protocol from [`NSCA`](https://exchange.nagios.org/directory/Addons/Passive-Checks/NSCA--2D-Nagios-Service-Check-Acceptor/details), the NSCA client used in this PR will not work. I'm not sure if there are any `NSCA-ng` clients written in Go, so you may have to write your own. See http://www.nsca-ng.org/documentation/protocol.html for the protocol information.

File: plugins/outputs/nsca/nsca.go
Comment (drunlade) [inline plugins/outputs/nsca/nsca.go]: I'm not sure I get why we call n.Close() immediately after connecting here. Can you add code comment to explain the behaviour?

File: plugins/outputs/nsca/nsca.go
Comment (drunlade) [inline plugins/outputs/nsca/nsca.go]: What's with the "E!" in error messages? Is this for your own development purposes or some sort of special required formatting?

File: plugins/serializers/nsca/nsca.go
Comment (drunlade) [inline plugins/serializers/nsca/nsca.go]: These should be configurable I would have thought?

Also, what if the machine has 8 cores?  How can this work on anything but a machine with 4 cores?

I would have thought CPU state warnings should be a % not a value of CPU cores. Take an input of 90%, check how many cores the machine has and work out the value to check against that way.

Am I missing something?

**Review by drunlade — COMMENTED** [review]
Made some comments

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go]: Oops, by mistake. will remove.

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go]: Yeah, its just for development purpose.

File: plugins/serializers/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/serializers/nsca/nsca.go]: Yeah, we can configure.

The values  CPU state we get from input plugin is not in percentage. 

I followed below link for Nagios CPU Load setup

https://serverfault.com/questions/209566/what-warning-and-critical-values-to-use-for-check-load

Comment (drunlade) [reply plugins/serializers/nsca/nsca.go]: Ah ok, so it's "config-per-server" rather than one global setup.

**Review by drunlade — APPROVED** [review]
Well, my issues are covered so Approving for now.

**Comment by gbjk** [comment]
This hasn't been merged?

File: plugins/outputs/nsca/nsca.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca.go]: These Comments in the sample config seem to be unrelated.

File: plugins/outputs/nsca/nsca.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca.go]: This shouldn't be hardcoded. or even like this.
Stub it with UUID, or panic, but this is too dangerous to be in there.

File: plugins/outputs/nsca/nsca.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca.go]: Isn't the identity meant to relate to the server?

File: plugins/outputs/nsca/nsca.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca.go]: UUID please.

File: plugins/outputs/nsca/nsca.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca.go]: I wouldn't expect a new serializer each time. Should be part of n, surely?

File: plugins/outputs/nsca/nsca.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca.go:115]: log.Error, Surely?

File: plugins/outputs/nsca/nsca_test.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca_test.go:14]: env var, or skip

**Comment by IrappaPattar** [comment]
@gbjk waiting for your approval.

File: plugins/outputs/nsca/nsca_test.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca_test.go]: This is from testify. It does not belong in here.

File: plugins/serializers/nsca/nsca.go
Comment (gbjk) [inline plugins/serializers/nsca/nsca.go]: It'd make more sense, and be more visible, if you just Sprintf this.

File: plugins/serializers/nsca/nsca.go
Comment (gbjk) [inline plugins/serializers/nsca/nsca.go:83]: switch case instead?

File: plugins/serializers/nsca/nsca.go
Comment (gbjk) [inline plugins/serializers/nsca/nsca.go]: Again, Sprintf

File: plugins/outputs/nsca/nsca.go
Comment (gbjk) [inline plugins/outputs/nsca/nsca.go:45]: Am I missing where this is used?

**Review by gbjk — CHANGES_REQUESTED** [review]

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go]: Removed

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go:45]: Its default method, which picks all the configured values.

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go]: Yeah, its relate to server.now  configured.

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go]: This is also server related and also configured.

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go]: Yeah, placed in connect method. where it performs once.

File: plugins/outputs/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca.go:115]: its just for debugging.

File: plugins/outputs/nsca/nsca_test.go
Comment (IrappaPattar) [reply plugins/outputs/nsca/nsca_test.go:14]: I didnt get your point.

File: plugins/serializers/nsca/nsca.go
Comment (IrappaPattar) [reply plugins/serializers/nsca/nsca.go:83]: I don't think we can use switch in this case.(if-else is best choice for less than & greater than cases)

**Comment by skldt** [comment]
@IrappaPattar could you please comment on the status of this PR?

Edit: This isn't urgent since it was just me commenting on older PRs that had me as a reviewer. Ignore this for now since I know you're working on high priority tasks currently and I don't want to distract you.

**Comment by IrappaPattar** [comment]
@skldt @gbjk could you please review this. due to some other prioritised tasks couldn't look at this PR.

**Comment by skldt** [comment]
@IrappaPattar sorry, I won't be able to look at this any time soon. There are other higher priority tasks now and it'll take a lot of time to get back onto this after so long. I doubt others would have the time now as well.

**Comment by skldt** [comment]
I think this can be closed without merging in. We haven't used this in a while and may not need to in the future.

@gbjk are you happy for this to be closed?

**Comment by gbjk** [comment]
God, SO irrellevant now!

**Pull request closed without merging**
